### PR TITLE
Performance improvements: Optimize built-in functions name checking and use `ConcurrentDictionary<string, WeakReference<LogicalExpression>>` at caching

### PR DESCRIPTION
**Pull request by gumbarros**
Before:
image

After:
(At IL it's optimized by string size)
image

### CODE REVIEW

**Comment by ThomasHambach** [comment]
Thanks for the contribution @gumbarros 

You mention performance improvements, did you run any tests to confirm it runs faster?

**Comment by gumbarros** [comment]
> Thanks for the contribution @gumbarros 
> 
> You mention performance improvements, did you run any tests to confirm it runs faster?

No, just checked the IL and checked the unit test of concurrency to see if everything is alright. Do you know how can I do a Benchmark to compare two NCalc versions?

**Comment by Bykiev** [comment]
> 
> Do you know how can I do a Benchmark to compare two NCalc versions?

You can use https://github.com/dotnet/BenchmarkDotNet library to measure performance with existing NuGet package

File: src/NCalc/Domain/EvaluationVisitor.cs [hunk 1]
```diff
@@ -276,333 +284,186 @@ public override void Visit(Function function)
...
-        {
-            #region Abs
-            case string n when n.Equals("abs", StringComparison.OrdinalIgnoreCase):
+        ExecuteBuiltInFunction(function);
+    }
 
-                CheckCase("Abs", function.Identifier.Name);
+    private void ExecuteBuiltInFunction(Function function)
+    {
+        var functionName = function.Identifier.Name.ToLower();
```
Comment (Bykiev) [inline src/NCalc/Domain/EvaluationVisitor.cs:292]: It'll be better to replace `ToLower()` with `ToUpperInvariant()`
You can find more discussions here: https://github.com/sklose/NCalc2/pull/84

**Comment by yallie** [comment]
I'm sorry guys, I'm still feeling sick a bit and can't function properly. 
I was going to merge all PRs and release it as a new version.
But looks like recent PRs have conflicts in the generated code due to grammar changes.
Could you give me a hand resolving the conflicts please?
Perhaps it's the matter of regenerating the parser/lexer classes.

PS. @gumbarros, @lucaschimweg — thanks a lot for your contributions! 🚀

**Comment by gumbarros** [comment]
Ready to merge

**Pull request merged**